docs/motivation.md

```diff
@@ -0,0 +1,550 @@
+reference capability `mut` as
+```verona
+type CMut = C & mut
+```
+
+By extension this means that polymorphism in Verona doesn't need special
+handling of capabilities in type parameters. A type `A[T]` parameterized on `T`
+can simply be instantiated with a type like `C & mut`. The method types of `A`
+are then able to express further restrictions on `T` that allow us to define
+operations for each instantiation.
+```verona
+type A[T] = {
+    withImm(v : T & imm) : A[T] & imm // immutable A[T] created with immutable T
+    withMut(v : T & mut) : A[T] & mut // mutable A[T] created with mutable T
+}
+```
+
+* TODO: compare with
+    - objective c: categories
+    - ruby: open classes
+    - C#: extension methods
+
+### Disjunction types
+Verona also includes disjunction types. Assuming two types `C` and `D` we can
+define the new type
+```verona
+type COrD = C | D
+```
+This describes the union of the two types.
+
+If we assume `C` and `D` above are classes, we can furthermore write pattern
+matches to distinguish between the two types dynamically.
+
+Since in Verona there is no subtyping between classes, we are able to check that
+a pattern match is exhaustive, and furthermore matching on concrete types will
+allow us to do implicit type filtering over match cases. E.g. imagine a case
+match like this
+```verona
+// C, D, E are classes
+
+f(x : C | D | E, g : (D | E) -> S) : S {
+    match x with {
+        (y : C) => { // y : C
+            ...
+        }
+        _ => { // x : D | E
+            g x
+        }
+    }
+}
+```
+We are able to implicitly filter the type of `x` in the default case.
+
+In a language like ML, the corresponding code would look like
+```ocaml
+type t = C of c
+       | D of d
+       | E of e
+
+
+f (x : t) (g : t -> s) =
+    match x with
+        C y => ...
+        _   => g x
+```
+Note that `g` cannot have a more refined type than `t -> s`.
+
+#### Type discrimination
+In ML we might want to construct a type such as
+```ocaml
+type s = A of int
+         B of int
+```
+A corresponding Verona type might look like
+```verona
+type S = I32 | I32
+```
+but it is now impossible to discriminate between the two cases. This is however
+easily fixed by defining `S` instead as
+```verona
+class A { var i : I32 }
+class B { var i : I32 }
+
+type S = A | B
+```
+
+#### Explicit nullability
+We can easily encode nullability with a class `None`
+```verona
+type A
+class None {}
+
+f(x : A | None) : S {
+    match x with {
+        (y: A) => { // y : A
+            ...
+        }
+        _ => { // x : None
+            ...
+        }
+    }
+}
+```
+
+### Subtyping and inheritance
+Verona is an object-oriented system with classes and traits. When it comes to
+subtyping, classes are treated nominally, and traits are treated structurally.
+Furthermore there is no subtyping between classes. This means a class
+can be a subtype of a trait, but never a subtype of a class other than itself.
```

Comment:
This can be covered early in a Structural section. Introduce `Eq` separately in a section that shows how to build useful structural primitives, and covers F-bound polymorphism and Self types.